r2dbc-core/src/main/java/com/google/cloud/sql/core/GcpConnectionFactoryProvider.java

```diff
@@ -0,0 +1,108 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import static io.r2dbc.spi.ConnectionFactoryOptions.Builder;
+import static io.r2dbc.spi.ConnectionFactoryOptions.DRIVER;
+import static io.r2dbc.spi.ConnectionFactoryOptions.HOST;
+import static io.r2dbc.spi.ConnectionFactoryOptions.PORT;
+import static io.r2dbc.spi.ConnectionFactoryOptions.PROTOCOL;
+
+import io.netty.handler.ssl.SslContextBuilder;
+import io.r2dbc.spi.ConnectionFactory;
+import io.r2dbc.spi.ConnectionFactoryOptions;
+import io.r2dbc.spi.ConnectionFactoryProvider;
+import java.util.Properties;
+import java.util.function.Function;
+
+/** {@link ConnectionFactoryProvider} for proxied access to GCP Postgres and MySQL instances. */
+public abstract class GcpConnectionFactoryProvider implements ConnectionFactoryProvider {
+
+  abstract ConnectionFactory tcpConnectonFactory(
+      Builder optionBuilder, Function<SslContextBuilder, SslContextBuilder> customizer);
+
+  abstract ConnectionFactory socketConnectionFactory(Builder optionBuilder, String socket);
+
+  abstract Builder createBuilder(ConnectionFactoryOptions connectionFactoryOptions);
+
+  abstract boolean supportedProtocol(String protocol);
+
+  @Override
+  public ConnectionFactory create(ConnectionFactoryOptions connectionFactoryOptions) {
+    String connectionName = connectionFactoryOptions.getRequiredValue(HOST);
+    String protocol = connectionFactoryOptions.getRequiredValue(PROTOCOL);
+
+    Properties properties = new Properties();
+    properties.put(CoreSocketFactory.CLOUD_SQL_INSTANCE_PROPERTY, connectionName);
+
+    if (supportedProtocol(protocol)) {
+      return createFactory(connectionFactoryOptions, properties);
+    }
+
+    throw new UnsupportedOperationException(
+        "Cannot create ConnectionFactory " + connectionFactoryOptions);
+  }
+
+  ConnectionFactory createFactory(
+      ConnectionFactoryOptions connectionFactoryOptions, Properties properties) {
+    Builder optionBuilder = createBuilder(connectionFactoryOptions);
+
+    // precompute SSL Data to avoid blocking calls during the connect phase.
+    CoreSocketFactory.getSslData(properties);
+    String socket = CoreSocketFactory.getUnixSocketArg(properties);
+
+    if (socket != null) {
+      return socketConnectionFactory(optionBuilder, socket);
+    }
+    return tcpConnectonFactory(optionBuilder, createSslCustomizer(optionBuilder, properties));
+  }
+
+  @Override
+  public boolean supports(ConnectionFactoryOptions connectionFactoryOptions) {
+    String driver = connectionFactoryOptions.getValue(DRIVER);
+    String protocol = connectionFactoryOptions.getValue(PROTOCOL);
+
+    return driver != null
+        && protocol != null
+        && driver.equals(getDriver())
+        && supportedProtocol(protocol);
+  }
+
+  @Override
+  public String getDriver() {
+    return "gcp";
```

Comment:
Does the driver really need to be different? Won't the driver be flexible?

Reply:
This is the prefix for the driver name, and each implementation would add its custom protocol name that follows that prefix. Users would use urls like this:
`r2dbc:gcp:postgres://<user>:<password>@<connection string>/<database>`
`r2dbc:gcp:mysql://<user>:<password>@<connection string>/<database>`